core/modules/mpls_pop.cc

```diff
@@ -0,0 +1,41 @@
+#include "mpls_pop.h"
+
+#include "../utils/ether.h"
+#include "../utils/mpls.h"
+
+
+void MPLSPop::ProcessBatch(bess::PacketBatch *batch) {
+  using bess::utils::be16_t;
+  using bess::utils::Ethernet;
+  using bess::utils::Mpls;
+
+  int cnt = batch->cnt();
+
+  for (int i = 0; i < cnt; i++) {
+    bess::Packet *pkt = batch->pkts()[i];
+
+    Ethernet *eth = pkt->head_data<Ethernet *>();
+
+    if (eth->ether_type != be16_t(Ethernet::Type::kMpls)) {
+      // Currently ignore non MPLS packets
+      continue;
+    }
+
+    // TODO(gsagie) save the MPLS label as metadata
+    // Mpls *mpls = reinterpret_cast<Mpls *>(eth_header + 1);
+
+    // TODO(gsagie) convert this to be more efficient using Intel instructions
+    Ethernet::Address src_addr = eth->src_addr;
+    Ethernet::Address dst_addr = eth->dst_addr;
+
+    pkt->adj(4);
+    Ethernet *eth_new = pkt->head_data<Ethernet *>();
+    eth_new->src_addr = src_addr;
+    eth_new->dst_addr = dst_addr;
+    eth_new->ether_type = be16_t(Ethernet::Type::kIpv4);
```

Comment:
Yes, probably configurable would be good. Config should allow to either specify the ethertype or, if bottom of stack, completely remove the ethernet encapsulation and expose the entire payload. The MPLS control plane knows what is going on. For example, if this label represents a pseudo wire, the right thing to do is to expose the entire MPLS payload as it represents a fully formed L2 packet.